unittests/test_pipeline.py

```diff
@@ -30,9 +28,25 @@ def _run(test, partition, prgenv):
     test.cleanup(remove_files=True)
 
 
-def load_test(testfile):
-    loader = RegressionCheckLoader(['unittests/resources/checks'])
-    return loader.load_from_file(testfile)
+@pytest.fixture
+def hellotest():
+    from unittests.resources.checks.hellocheck import HelloTest
+    yield HelloTest
```

Comment:
> because the vars are inserted as descriptors, they become class attributes after the first class instantiation.

Actually, it's because they are inserted as descriptors too late, at the final object instantiation, which is also something you wouldn't expect for a descriptor. Couldn't they be inserted earlier, during the class creation? Is it just for uniformity with parameters? Parameters needed to be injected at object creation, but why the variable descriptors? 

This behaviour is *really* counter-intuitive and I'm afraid we will have to face it the hard way sometime in the future.

I've also seen the following "inconsistency". If I do the following in the `HelloTest`

```python
@rfm.simple_test
class HelloTest(rfm.RegressionTest):
    var('name', int)

```

I correctly get the `re-declaration` error, even if I disable the sanity checking. However, this is not case for other variables, e.g., `prefix`, `current_environ` etc. (which is the expected behaviour with the current implementation if sanity checking is disabled).

Reply:
> Couldn't they be inserted earlier, during the class creation? 

This would bring us back to square 1. 
The reason why the variables get inserted at the very end boils down to getting more control on the vars and their interaction with the user. This is unrelated with whether they're descriptors or not. If you insert the vars during class creation, a user could do
```python
class Foo(rfm.RegressionTest):
    num_tasks = 'bananas'
    def __init__(self):
        ...
```
and that's enough to break everything. If you insert the variables at the end, you know that the user has messed up and then you can throw the right error. There are now unit tests in `test_variables.py` to test for an error when overriding a builtin variable (`name`) and builtin attributes (`current_environ`).

The "instantiate and inherit" issue you pointed out above has now been fixed by making the variable space a bit smarter. When the variables get injected into the class, the variable space now tracks which variables have been injected, and it knows there can be no conflict with those variables at any point onwards.